### PR TITLE
Add multi-GPU support with `accelerate`

### PR DESCRIPTION
**Pull request by GavEdwards**
# Summary
 
Enable GPU support (+more) via the Accelerate library. 

This is still work in progress - there's still some bugs to be ironed out around multi-gpu and some models. 

TODO:
- [ ] add documentation for accelerate
- [ ] multi-gpu tests
- [ ] test with all models 

--- 
- [ ] Unit tests provided for these changes
- [ ] Documentation and docstrings added for these changes using the [sphinx style](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html)

## Changes 

* Add Accelerate as a dependency 
* Adjust the pipeline code to use accelerate

### CODE REVIEW

**Comment by benedekrozemberczki** [comment]
Packedgraphs do not have a .to() method, but a custom .cuda() method. Pretty messed up.

https://github.com/DeepGraphLearning/torchdrug/blob/master/torchdrug/data/graph.py

**Comment by GavEdwards** [comment]
Examples: 
https://github.com/huggingface/accelerate/tree/main/examples

**Comment by cthoyt** [comment]
So one main thing about this PR is that I don't think it's necessary to use Accelerate to add GPU support - we only have to more cleverly keep track of devices. I'm not against using Accelerate since it has some nice add-ons for multi-gpu etc. but I wouldn't depend on it to solve the original problem

Perhaps we can monkey patch a to() into the packed graph class

**Comment by GavEdwards** [comment]
> So one main thing about this PR is that I don't think it's necessary to use Accelerate to add GPU support - we only have to more cleverly keep track of devices. I'm not against using Accelerate since it has some nice add-ons for multi-gpu etc. but I wouldn't depend on it to solve the original problem
> 
> Perhaps we can monkey patch a to() into the packed graph class

Hi @cthoyt 😄  My thinking was this approach avoids us having to re-invent the wheel and quickly solves the current gpu need. The original issue doesn't describe requirements to aim for #65.

Two questions: 
* In your opinion, what parts of the original problem doesn't Accelerate solve?
* What requirements need to be met for this approach be accepted? 

I'm not attached to Accelerate as a solution either, just trying to understand limits & needs

**Comment by cthoyt** [comment]
I'm saying that implementing GPU usability and implementing `accelerate` are two independent things, and I would rather see a PR that first enables GPU usage explicitly *without* `accelerate` to make sure we don't make any `accelerate`-specific mistakes

File: chemicalx/pipeline.py [hunk 1]
```diff
@@ -152,10 +184,12 @@ def pipeline(
     for _epoch in trange(epochs):
         for batch in train_generator:
             optimizer.zero_grad()
-            prediction = model(*model.unpack(batch))
-            loss_value = loss(prediction, batch.labels)
+
+            device_batch = to_device(model.unpack(batch), device)
```
Comment (cthoyt) [inline chemicalx/pipeline.py:188]: I think the batch generator should know what device to put the batches on so this doesn't have to be changed in the pipeline

Comment (cthoyt) [reply chemicalx/pipeline.py:188]: I.e., the `BatchGenerator.__init__` should take an optional `torch.device` (if not given, assume CPU) and the generation steps should take care of moving the tensors over to the appropriate device

**Comment by cthoyt** [comment]
@GavEdwards please note we've already merged a simple solution into the main branch and now updated your PR with it, please check it out

File: setup.py [hunk 2]
```diff
@@ -15,6 +15,9 @@
     "pystow",
     "pytdc",
     "more-itertools",
+    "accelerate",
+    # FIXME what is packaging for?
+    "packaging",
```
Comment (cthoyt) [inline setup.py:20]: what's `packaging` for?

File: chemicalx/pipeline.py [hunk 3]
```diff
@@ -70,6 +71,45 @@ def save(self, directory: Union[str, Path]) -> None:
         )
 
 
+def to_device(objects, device):
```
Comment (cthoyt) [inline chemicalx/pipeline.py:74]: I don't think this is necessary now that #84 is merged

Comment (GavEdwards) [reply chemicalx/pipeline.py:74]: Agreed, will close the PR

Comment (cthoyt) [reply chemicalx/pipeline.py:74]: @GavEdwards I was only referring to the `to_device` function. There's still some benefit to consider adding accelerate for multi-GPU or TPU usage

**Pull request closed without merging**